js/src/framework2.ts

```diff
@@ -44,6 +62,51 @@ export class Project {
     this.prompts = new PromptBuilder(this);
     this.scorers = new ScorerBuilder(this);
   }
+
+  public addPrompt(prompt: CodePrompt) {
+    this._publishablePrompts.push(prompt);
+    if (globalThis._lazy_load) {
+      globalThis._evals.prompts.push(prompt);
+    }
+  }
+
+  public addCodeFunction(
+    fn: CodeFunction<
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      any,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      any,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      GenericFunction<any, any>
+    >,
+  ) {
+    this._publishableCodeFunctions.push(fn);
+    if (globalThis._lazy_load) {
+      globalThis._evals.functions.push(fn);
+    }
+  }
+
+  async publish() {
+    await login();
+    const projectMap = new ProjectNameIdMap();
+    const functionDefinitions: FunctionEvent[] = [];
+    if (this._publishableCodeFunctions.length > 0) {
+      throw new Error(
```

Comment:
I wonder if we should warn rather than throw an exception. I could see it annoying since it's all or nothing. i.e. if I have a .create() for functions and prompts, then eventually I hit .publish() nothing will be pushed. Which is safer, but I don't think we got anything out of this change. I would have to split up the file and avoid calling .create() for functions.

Reply:
sounds good to me